sled-agent/src/illumos/vnic.rs

```diff
@@ -92,12 +117,17 @@ impl VnicAllocator {
 pub struct Vnic {
     name: String,
     deleted: bool,
+    kind: VnicKind,
 }
 
 impl Vnic {
     /// Takes ownership of an existing VNIC.
     pub fn wrap_existing<S: AsRef<str>>(name: S) -> Self {
-        Vnic { name: name.as_ref().to_owned(), deleted: false }
+        Vnic {
+            name: name.as_ref().to_owned(),
+            deleted: false,
+            kind: VnicKind::from_name(name.as_ref()),
+        }
     }
 
     /// Deletes a NIC (if it has not already been deleted).
```

Comment:
I'm looking at your implementation of `delete_all_xde_devices` - it looks like the implementation of `delete_vnic` might need to be different if this is an XDE device. Should we be matching on `kind` in this `delete` implementation?

Reply:
There are two data links here: the xde device, and the VNIC over that. The latter can be deleted via `dladm`. The former must be deleted by an `ioctl(2)` call to the xde driver. It might be possible to put those together, by making a call into a hypothetical `sled_agent::opte::delete_xde_device` inside the `Vnic::delete` call. That hypothetical function would just be the singular, non-looping form of `sled_agent::opte::delete_all_xde_devices`.

Reply:
Gotcha - okay, so this current implementation isn't wrong, it's basically just deleting "one half" of the device.

I'll mark this conversation resolved, if that use-case is acceptable!